### ISSUE TITLE
TextInput - StageText/set viewPort ArgumentError #2004:

### ISSUE THREAD

**Issue by teotigraphix** [BowlerHatLLC/feathers#1646]
I had completely working code and something changed in Feathers because I have not touched this app code since 3.1-3.2

What could this possibly be, I understand the height to be like -14 going into the viewPort, that is the actual error.

```
Main Thread (Suspended: ArgumentError: Error #2004: One of the parameters is invalid.)	
	flash.text::StageText/set viewPort [no source]	
	feathers.controls.text::StageTextTextEditor/refreshViewPortAndFontSize	
	feathers.controls.text::StageTextTextEditor/layout	
	feathers.controls.text::StageTextTextEditor/draw	
	feathers.core::FeathersControl/validate	
	feathers.controls::TextInput/layoutChildren	
	feathers.controls::TextInput/draw	
	feathers.core::FeathersControl/validate	
	feathers.layout::VerticalLayout/validateItems	
	feathers.layout::VerticalLayout/layout	
	feathers.controls::LayoutGroup/draw	
	com.teotigraphix.ui.component::LabeledTextInput/draw	
	feathers.core::FeathersControl/validate	
	feathers.controls::LayoutGroup/validate	
	feathers.layout::VerticalLayout/validateItems	
	feathers.layout::VerticalLayout/layout	
	feathers.controls::LayoutGroup/draw	
	feathers.core::FeathersControl/validate	
	feathers.controls::LayoutGroup/validate	
	feathers.controls::Scroller/refreshViewPortBoundsForMeasurement	
	feathers.controls::Scroller/refreshViewPort	
	feathers.controls::Scroller/draw	
	feathers.controls::ScrollContainer/draw	
	feathers.controls::Panel/draw	
	caustic.guide.ui.screen.export.drawers::ExportSoundsGlobalScreen/draw	
	feathers.core::FeathersControl/validate	
	feathers.controls::ScrollContainer/validate	
	feathers.controls.supportClasses::BaseScreenNavigator/layoutChildren	
	feathers.controls.supportClasses::BaseScreenNavigator/draw	
	feathers.core::FeathersControl/validate	
	feathers.controls.supportClasses::BaseScreenNavigator/showScreenInternal	
	feathers.controls::StackScreenNavigator/pushScreen	
	com.teotigraphix.ui.app.core::ScreenLauncherImpl/pushRightScreen	
	com.teotigraphix.ui.app.core::ScreenLauncherImpl/setScreenState	
	caustic.guide.ui.screen.export._mediators::ExportAsSoundsScreenMediator/refreshView	
	com.teotigraphix.ui.app::XAbstractMediator/initializeView	
	com.teotigraphix.ui.app::XAbstractMediator/onRegister	
	org.robotlegs.starling.base::MediatorBase/onCreationComplete	
	starling.events::EventDispatcher/invokeEvent	
	starling.events::EventDispatcher/dispatchEvent	
	starling.display::DisplayObject/dispatchEvent	
	starling.events::EventDispatcher/dispatchEventWith	
	feathers.core::FeathersControl/validate	
	feathers.controls::ScrollContainer/validate	
	feathers.controls.supportClasses::BaseScreenNavigator/layoutChildren	
	feathers.controls.supportClasses::BaseScreenNavigator/draw	
	feathers.core::FeathersControl/validate	
	feathers.core::ValidationQueue/advanceTime	
	feathers.controls.supportClasses::BaseScreenNavigator/showScreenInternal	
	feathers.controls::ScreenNavigator/showScreen	
	com.teotigraphix.ui.app.core::ScreenLauncherImpl/setScreen	
	com.teotigraphix.app.controller.handlers.overridable::OnApplicationNavigationItemChangedHandler/onNavigationItemChanged	
	org.osflash.signals::Signal/dispatch	
	com.teotigraphix.ui.app.core::AbstractUIState/set selectedNavigationItem	
	com.teotigraphix.ui.drawers.navigation._mediators::CollapsibleListPanelMediator/view_changedHandler	
	org.robotlegs.starling.base::EventMap/routeEventToListener	
	Function/<anonymous>	
	starling.events::EventDispatcher/invokeEvent	
	starling.events::EventDispatcher/bubbleEvent	
	starling.events::EventDispatcher/dispatchEvent	
	starling.display::DisplayObject/dispatchEvent	
	starling.events::EventDispatcher/dispatchEventWith	
	com.teotigraphix.ui.drawers.navigation::CollapsibleListPanel/list_changeHandler	
	com.teotigraphix.ui.drawers.navigation::CollapsibleListPanel/__list_change	
	starling.events::EventDispatcher/invokeEvent	
	starling.events::EventDispatcher/dispatchEvent	
	starling.display::DisplayObject/dispatchEvent	
	starling.events::EventDispatcher/dispatchEventWith	
	feathers.controls::List/selectedIndices_changeHandler	
	starling.events::EventDispatcher/invokeEvent	
	starling.events::EventDispatcher/dispatchEvent	
	starling.events::EventDispatcher/dispatchEventWith	
	feathers.data::ListCollection/set data	
	feathers.controls.supportClasses::ListDataViewPort/renderer_changeHandler	
	starling.events::EventDispatcher/invokeEvent	
	starling.events::EventDispatcher/dispatchEvent	
	starling.display::DisplayObject/dispatchEvent	
	starling.events::EventDispatcher/dispatchEventWith	
	feathers.controls::ToggleButton/set isSelected	
	feathers.utils.touch::TapToSelect/target_touchHandler	
	starling.events::EventDispatcher/invokeEvent	
	starling.events::TouchEvent/dispatch	
	starling.events::Touch/dispatchEvent	
	starling.events::TouchProcessor/processTouches	
	starling.events::TouchProcessor/advanceTime	
	starling.core::Starling/advanceTime	
	starling.core::Starling/nextFrame	
	starling.core::Starling/onEnterFrame	
```

**Comment by teotigraphix** [comment]
This is the variables of stageTextViewPort	

```
stageTextViewPort	Rectangle (@10426ee1)	
	bottom	17 [0x11]	
	bottomRight	Point (@103a0561)	
	height	19 [0x13]	
	left	NaN	
	right	NaN	
	size	Point (@103a0581)	
	top	-2 [0xfffffffe]	
	topLeft	Point (@103a05a1)	
	width	976 [0x3d0]	
	x	NaN	
	y	-2 [0xfffffffe]	

```

**Comment by teotigraphix** [comment]
And the starlingViewPort	

```
starlingViewPort	Rectangle (@3676a91)	
	bottom	611 [0x263]	
	bottomRight	Point (@103a0261)	
	height	611 [0x263]	
	left	0	
	right	1008 [0x3f0]	
	size	Point (@103a0281)	
	top	0	
	topLeft	Point (@103a0361)	
	width	1008 [0x3f0]	
	x	0	
	y	0	

```

**Comment by teotigraphix** [comment]
matrix and point

```
matrix	Matrix (@102aed81)	
	a	1	
	b	0	
	c	0	
	d	1	
	tx	NaN	
	ty	0	
point	Point (@1af294e1)	
	length	NaN	
	x	NaN	
	y	-2 [0xfffffffe]	
```

**Comment by joshtynjala** [comment]
Thanks, I'll investigate. I'm guessing that those NaN values are what's causing the runtime error, but I'll need to figure out where they're coming from.

> I understand the height to be like -14 going into the viewPort

That's interesting. Are you setting the height of the TextInput to a negative value on purpose?

**Comment by teotigraphix** [comment]
Well I am not setting it to -14. This is a custom component that holds the TextInput, it is like  labeled TextInput with prompt. So there are 3 components in a vertical layout that the TextInput is inside.

I have a feeling it just has to do with you changing something in layout code and my older code is not treating invalidation or the bounds correctly.

I wrote that component about 1.5 years ago.

Probably since I need to update this app I will just rewrite the whole screen, maybe I can at leaset tell you where the error originated.

**Comment by teotigraphix** [comment]
BTW, the last time I released this app was around June 20th, I think the framework was 3.2 might have been 3.1(if it was released then).

I was pretty regular keeping up with you in the 3.x branch.

**Comment by teotigraphix** [comment]
I figure out where the NaN is coming from;

The call `StageTextEditor.refreshViewPortAndFontSize()`

`this.getTransformationMatrix(this.stage, matrix);`

When `matrix` goes into that method, tx comes out NaN.

**Comment by teotigraphix** [comment]
Actually, I think I found the culprit.

Stepping through, I realized in the concat() loop, that as it was cycling through parents, my RightDrawer tx is NaN, so it looks like this is coming from a Drawer in the root DrawersApplication.

So I will try to debug what is going on in the drawer that is messed up on size.

**Comment by teotigraphix** [comment]
Ok, looking at the Drawer's x, it is NaN.

Can a Drawer's x be NaN in your layout code?

height, width and y are all fine, it's just the x NaN and the drawer is visible = false.

**Comment by joshtynjala** [comment]
Feathers would never set the x property to NaN on purpose. It must be propagating from somewhere else.

**Comment by teotigraphix** [comment]
I found it.

Drawers.layoutChildren() at 3141.

`this._rightDrawer.x = rightDrawerX;

rightDrawerWidth is Infinity.

**Comment by joshtynjala** [comment]
Now, the question becomes: why would the width of the drawer be set to infinity?

**Comment by teotigraphix** [comment]
Damn, this was my error when I changed something last month. I do some weird sizing with drawers using the maxWidth of the child. The drawer is a stack.

So some drawers size to stage but some use a maxWidth I set.

I forgot to add this line when I added a branch statement.

` if (IFeathersControl(rootScreen).maxWidth != Number.POSITIVE_INFINITY)`

Now I have 

`
            var rootScreen:IScreen = rightDrawer.popToRootScreen() as IScreen;
            if (IFeathersControl(rootScreen).maxWidth != Number.POSITIVE_INFINITY)
                rightDrawer.width = IFeathersControl(rootScreen).maxWidth;
`

All works again. 

That was probably a bad hack because I made this mistake but it worked fine until I added another line that was just;

`
rightDrawer.width = IFeathersControl(rootScreen).maxWidth;
`
Since I didn't explicitly define a max width in my screen, it was setting the width to infinity.

This was an older app, my newer ones, the right drawer is usually 60 wide so I didn't run into this.
